[PXT-765] Support for Office Formats as part of Document Type through MarkdownIT

pierrebrunelle opened the pull request:
## Add PowerPoint, Word, and Excel support via MarkItDown

**Core Integration:**
- Added `markitdown[pptx,docx,xlsx]>=0.1.3` dependency
- Extended `DocumentType.DocumentFormat` enum with `PPTX`, `DOCX`, `XLSX`
- Created `get_office_handle()` that converts office files to markdown AST via MarkItDown
- Added `_markitdown_sections()` to `DocumentSplitter` for processing converted documents
- Registered `markitdown` as optional package in `env.py`

**Design Decision:**
Office documents are converted to markdown on load and processed through the existing markdown AST pipeline. This provides consistent behavior across all document types and leverages existing splitting/metadata logic.

**Bug Fix:**
Fixed edge case in `_token_chunks()` where Unicode decode failures could cause assertion errors on certain document content.

**Testing:**
- Added 3 tests in `test_document.py` covering insert, handle creation, and splitting
- Tested with 26,971-line CalPy presentation (splits into 2,369 token chunks)

**Documentation:**
- Created cookbook recipe: `doc-extract-text-from-office-files.ipynb`
- Updated cookbook README, CONTRIBUTING.md, and STYLE_GUIDE.md

### Supported Formats
- PowerPoint: `.pptx`, `.ppt`
- Word: `.docx`, `.doc`  
- Excel: `.xlsx`, `.xls`

aaron-siegel commented on pixeltable/iterators/document.py:
```diff
@@ -472,6 +480,67 @@ def _txt_sections(self) -> Iterator[DocumentSection]:
         assert self._doc_handle.txt_doc is not None
         yield DocumentSection(text=ftfy.fix_text(self._doc_handle.txt_doc), metadata=DocumentSectionMetadata())
 
+    def _markitdown_sections(self) -> Iterator[DocumentSection]:
```
This looks like pure slop to me. Why does this need to be a separate method from `_markdown_sections()`? Either reuse `_markdown_sections()` and parameterize as necessary to minimize code duplication, or add a comment to explain the reasoning why this needs to be separate.

pierrebrunelle commented on tests/test_document.py:
```diff
@@ -47,6 +47,17 @@ def valid_doc_paths(self) -> list[str]:
+
+        docs_dir = Path(__file__).parent / 'data' / 'documents'
+        office_files = []
+        for ext in ['.pptx', '.docx', '.xlsx']:
```
Done.

pierrebrunelle commented on tests/test_document.py:
```diff
@@ -47,6 +47,17 @@ def valid_doc_paths(self) -> list[str]:
+        docs_dir = Path(__file__).parent / 'data' / 'documents'
+        office_files = []
+        for ext in ['.pptx', '.docx', '.xlsx']:
+            for f in docs_dir.glob(f'*{ext}'):
```
Done. Simplified to single list comprehension.

pierrebrunelle commented on pixeltable/utils/documents.py:
```diff
@@ -1,7 +1,7 @@
 import os
 
-import bs4
+import bs4  # type: ignore[import-untyped]
```
Removed. Unnecessary since we have types-beautifulsoup4 installed.

aaron-siegel commented on tests/test_document.py:
```diff
@@ -47,6 +47,13 @@ def valid_doc_paths(self) -> list[str]:
 
+    def office_doc_paths(self) -> list[str]:
+        """Return paths to office format test documents."""
+        from pathlib import Path
```
Move to top of file (there is no justification for an inline import here).

pierrebrunelle replied:
Done.

aaron-siegel commented on tests/test_document.py:
```diff
@@ -47,6 +47,13 @@ def valid_doc_paths(self) -> list[str]:
     def invalid_doc_paths(self) -> list[str]:
         return [get_video_files()[0], get_audio_files()[0], get_image_files()[0]]
 
+    def office_doc_paths(self) -> list[str]:
```
Missing are test documents from all supported formats.

pierrebrunelle commented on pixeltable/iterators/document.py:
```diff
@@ -365,9 +373,16 @@ def process_element(el: Tag | NavigableString) -> Iterator[DocumentSection]:
-    def _markdown_sections(self) -> Iterator[DocumentSection]:
-        """Create DocumentSections reflecting the html-specific separators"""
-        assert self._doc_handle.md_ast is not None
+    def _markdown_sections(self, md_ast: list[dict] | None = None) -> Iterator[DocumentSection]:
```
Done. md_ast is now a required parameter.

pierrebrunelle commented on pixeltable/iterators/document.py:
```diff
@@ -365,9 +373,16 @@ def process_element(el: Tag | NavigableString) -> Iterator[DocumentSection]:
+            md_ast: The markdown AST to process. If None, uses self._doc_handle.md_ast.
+                    This allows reuse for office formats converted to markdown via MarkItDown.
+        """
+        if md_ast is None:
```
Done.

aaron-siegel commented on pixeltable/utils/documents.py:
```diff
@@ -93,3 +99,21 @@ def get_txt(path: str) -> str:
+    # Parse the markdown into an AST for consistent handling
+    md_parser = mistune.create_markdown(renderer=None)
+    # When renderer=None, mistune returns the AST as a list
+    return md_parser(markdown_text)  # type: ignore[return-value]
```
Same, why not just use the correct return type?

aaron-siegel commented on tests/test_document.py:
```diff
@@ -47,6 +48,15 @@ def valid_doc_paths(self) -> list[str]:
     def invalid_doc_paths(self) -> list[str]:
         return [get_video_files()[0], get_audio_files()[0], get_image_files()[0]]
 
+    def office_doc_paths(self) -> list[str]:
```
Next question :) Why is this special-cased rather than just using the existing `utils.get_documents()` (which already picks up everything in tests/documents) and incorporating into the existing document tests? (Also see the below comment)

pierrebrunelle commented:
Addressed all review comments:

**documents.py type annotations**: Changed `# type: ignore[return-value]` to proper `cast(list, ...)` with explanatory comments. The issue is that mistune's `create_markdown(renderer=None)` is typed to return `str | list` but actually returns `list[dict]` when renderer is None.

**Redundant office tests**: Removed the separate `office_doc_paths()` method and the three office-specific test functions (`test_office_formats_insert`, `test_office_formats_get_handle`, `test_office_formats_splitter`). The office test documents are already in `tests/data/documents/` and are picked up by `get_documents()`, so `test_doc_splitter` now covers them. Added `skip_test_if_not_installed('markitdown')` to handle environments without the package.

pierrebrunelle reviewed (commented):
Done.

pierrebrunelle commented on pixeltable/utils/documents.py:
```diff
@@ -61,14 +67,14 @@ def get_html_handle(path: str) -> bs4.BeautifulSoup:
         text = file.read()
     md_ast = mistune.create_markdown(renderer=None)
-    return md_ast(text)
+    return md_ast(text)  # type: ignore[return-value]
```
Done. Changed to `cast(list, ...)` with a comment explaining that mistune types it as `str | list` but actually returns `list[dict]` when `renderer=None`.

pierrebrunelle commented on tests/test_document.py:
```diff
@@ -372,3 +385,93 @@ def test_doc_splitter_images(self, reset_db: None) -> None:
+                assert handle.format == ts.DocumentType.DocumentFormat.XLSX, path
+                assert handle.md_ast is not None, path
+
+    def test_office_formats_splitter(self, reset_db: None) -> None:
```
Removed. The office test documents are in `tests/data/documents/` and picked up by `get_documents()`, so `test_doc_splitter` handles them. Added `skip_test_if_not_installed('markitdown')` to handle environments without the package.

aaron-siegel commented on pixeltable/utils/documents.py:
```diff
@@ -93,3 +101,19 @@ def get_txt(path: str) -> str:
+    markdown_text = result.text_content
+
+    md_parser = mistune.create_markdown(renderer=None)
+    # renderer=None returns AST as list[dict], but mistune types it as str | list
```
You can also do it this way (usually preferred to using `cast()`):
```
result = md_parser(markdown_text)
assert isinstance(result, list)
return result
```
and mypy will understand how to interpret it.

pierrebrunelle replied:
done
